### ISSUE TITLE
Negative values for CD loss

### ISSUE THREAD

**Issue by nithishdivakar** [wuaalb/keras_extensions#7]
In the implementation of Contrastive Divergence loss [see here](https://github.com/wuaalb/keras_extensions/blob/master/keras_extensions/rbm.py#L218), What is stopping the value to be negative...? A loss functions value should not be negative...

I'm doing a training and I'm getting negative values of loss. This is causing the model to give bad prediction eventually.

**Comment by wuaalb** [comment]
I don't think there's any theoretical constraint why the loss function should not be negative; it is just a function that is being minimized..

That said, I think it should normally be positive, so maybe you're using a learning rate that's too high?

**Comment by nithishdivakar** [comment]
I'm using GB-RBMs
How should I scale data? Is mean centering batch wise is enough?
In the test phase, the values I'm getting are more or less binary even though I've used Gaussian visible layers.

P.S: is it possible to add an example of GBRBMs  in which they generate some data out of visible layer eventually?

**Comment by wuaalb** [comment]
I usually pre-process data to have zero mean and unit variance.

In test phase you're doing reconstruction (x -> h -> x_rec)? Or generating random h?
I've used the GB-RBM for audio and it worked as expected.

Proper examples would be great (see #3); but I don't really have time to do them.

**Comment by nithishdivakar** [comment]
Im using following coode to create inference model and sampling.
```python
for i in xrange(5):
     inference_model.add(rbm.get_h_given_x_layer(as_initial_layer=(i==0)))
     inference_model.add(SampleBernoulli(mode='maximum_likelihood'))
     inference_model.add(rbm.get_x_given_h_layer())

:
:
hat_x = inference_model.predict(x)
```

**Comment by wuaalb** [comment]
What's inside the for loop looks OK, but the loop itself doesn't make sense I think.
Could you try with just a single GB-RBM?

A 2-layer DBN should be something like
```
   x -> GB-RBM -> h1 -> RBM -> h2 -> RBM -> ~h1 -> GB-RBM -> ~x
```
But I've never tried this..

**Comment by nithishdivakar** [comment]
Can you sketch a small outline on how to train a DBN with the current settings?


The loop was for making the model generate sample after 5 gibbs sampling.
